### DISCUSSION TITLE
Preventing mutation of a function result which should be immutable

### DISCUSSION THREAD

**Discussion by ntrel** [vlang/v#13812] [Ideas]
Issue https://github.com/vlang/v/issues/13791 shows that assigning an immutable slice to a mutable variable should be disallowed:
```v
arr := [1, 2, 3, 4, 5]
mut s := arr[1..2] // should error
s[0] = 0
```
That is probably straightforward to implement. But what about when the immutable slice is the result of a function call:
```v
fn id(arr []int) []int { return arr[1..2] }
arr := [1, 2, 3, 4, 5]
mut s := id(arr) // should error
s[0] = 0
```
I think we need to allow marking the return type as `mut`. This could maybe be done with `vfix`. Then any mutation of a function result which doesn't have that `mut` should be made an error. This will also need tracking of every expression as `mut` or not:
```v
fn id(arr []int) []int { return arr[1..2] }
arr := [1, 2, 3, 4, 5]
id(arr)[0] = 0 // should error
```

**Comment by CoderCharmander** [comment]
In my opinion, mutability for slices could be a part of the type. For example, `[]mut int` would be able to be mutated, but `[]int` not. This could also be extended for taking slices, for example:
```vlang
arr := [1, 2, 3]
// s := mut arr[1..2] // illegal because arr is immutable
s := arr[1..2]
// s[0] = 0 // illegal because s is not a mutable slice
```
This could be useful in situations like this, and it can be added before the 1.0 syntax freeze.

However, the problems with this approach are that arrays are currently the same type as slices, and putting mutability in the types of "owned" arrays (that are not a slice of anything) (borrowed Rust terminology) does not really make sense; instead, it should be a property of the variable owning the array.

**Comment by Wajinn** [comment]
I don't think the problem is assigning an immutable slice to a mutable variable.  Rather, the problem is assigning an immutable slice to a mutable variable, **and then** changes made to the mutable variable shows up in the immutable variable.  So the below, is a problem.

```
arr := [1, 2, 3, 4, 5]
mut s := arr[1..2]
s[0] = 20
println(arr)
[1, 20, 3, 4, 5]
```

As long as changes to the contents of the mutable variable are not showing up in the immutable variable, then I would think that's what is usually wanted.  The mutable variable should be a copy to do whatever, and not affecting the original immutable.

**Reply by CoderCharmander** [reply]
Making a clone implicitly might be a performance problem and the needed behavior can already be reached with `arr[1..2].clone()`.

**Reply by CoderCharmander** [reply]
Also, if I added a `mut` somewhere to a previously immutable variable, I would not expect the behavior or performance of the code to change (except for a compiler warning about an useless `mut`)
